### ISSUE TITLE
Bug: useSyncExternalStore does not schedule update after mutation

### ISSUE THREAD

**Issue by bvaughn** [facebook/react#27670]
Here is a Replay recording with comments:
https://app.replay.io/recording/318114a4-3102-4732-ad1b-cb63b9c0ac22

I think the comments I've added show the following scenario:
1. A React component (`Subscriber`) uses `useSyncExternalStore` to read from a mutable store.
2. The Activity subtree containing `Subscriber` is hidden, and React unsubscribes from the store.
3. An update is scheduled to show the hidden subtree again and the `Subscriber` component re-renders (with memoized state).
4. A parent component mutates the store in a layout effect. (The `Subscriber` component is not yet listening and so it does not re-render to reflect the updated store value.)
6. React re-subscribes the `Subscriber` component (`useSyncExternalStore`) but it has already missed the mutation **and React does not check for a changed snapshot value**.

We originally observed this behavior in Replay itself but I was able to reduce it to the following simplified case:
https://codesandbox.io/s/inspiring-bird-m4wv5l

I've added comments to the Replay, including ones that bracket the problematic `commitRoot`. Here's a short Loom as well talking through the bug:
https://www.loom.com/share/2584cad5b4c44e6bba396ff8cf79db1d

I _think_ an application could work around this issue in a couple of ways:
* Only mutate stores in passive effects. (This may cause visible layout shift though so it seems bad.)
* Mutable store users should avoid memoized `getSnapshot` and `subscribe` functions. (This would schedule some unnecessary effects work which is probably nice to avoid but maybe acceptable.)
* **Store subscribe methods should always invoke the `callback` React passes.** (This would cause a lot of unnecessary `Object.is` comparisons but that's probably an acceptable cost.)

I think the third option above seems best, but needing to do that feels like a foot gun for the API. Thoughts?

cc @acdlite in case you find this interesting

**Comment by bvaughn** [comment]
If it would be useful, I can create a failing unit test from my Code Sandbox. (This is probably pretty straight forward to do though FWIW, so I imagine someone else could do it as well.)

**Comment by sophiebits** [comment]
If I’m not mistaken, this line is meant to catch if the store was mutated in between render and subscribe (if we were not already subscribed to the store at render time):

https://github.com/facebook/react/blob/746890329452cbec8685eb3466847c5f17d9dc77/packages/react-reconciler/src/ReactFiberHooks.js#L1689

Perhaps this code isn’t running at the right time when reviving the tree.

**Comment by bvaughn** [comment]
Agreed. That code only runs during mount, though, or if the getSnapshot function identity has changed. In this case it doesn’t get run bc the hidden to visible transition is an update and getSnapshot is a stable function reference.

**Comment by rickhanlonii** [comment]
@acdlite are we missing a mutated value check here?

**Comment by bvaughn** [comment]
Friendly ping.

**Comment by eps1lon** [comment]
This is only a bug when used with `Activity` or does it also happen when a suspended tree unsuspends? 

If it's just with `Activity` it may take a while to fix it since it's currently being deprioritized: https://react.dev/blog/2024/02/15/react-labs-what-we-have-been-working-on-february-2024#offscreen-renamed-to-activity

**Comment by bvaughn** [comment]
@eps1lon I already wrote a test for this in PR #27670

**Comment by bvaughn** [comment]
> This is only a bug when used with `Activity` or does it also happen when a suspended tree unsuspends?

The scope of the bug I've reported here requires Offscreen/Activity.

> I'm still trying to write a test for it so that we can pick it up again at a later stage.

@eps1lon I already wrote a test for this in PR #27676

**Comment by eps1lon** [comment]
You can also repro this when you remove the `store.set` from the parent and move it into a passive effect **before** the `useSyncExternalStore` call:
```jsx
    function App({mode, revision}) {
      return (
        <React.unstable_Activity mode={mode}>
          <Subscriber revision={revision} />
        </React.unstable_Activity>
      );
    }

    function Subscriber({revision: propRevision}) {
      React.useEffect(() => {
        console.log('store.set');
        store.set('revision:' + propRevision);
      }, [propRevision]);
      const revision = useSyncExternalStore(store.subscribe, store.getState);
      return <Text text={revision} />;
    }
```
 If that passive effect is moved **after** the `useSyncExternalStore`, it passes again because then React will have subscribed to store changes before the passive effect runs
```diff
-React.useEffect(() => {
-  console.log('store.set');
-  store.set('revision:' + propRevision);
-}, [propRevision]);
 const revision = useSyncExternalStore(store.subscribe, store.getState);
+React.useEffect(() => {
+  console.log('store.set');
+  store.set('revision:' + propRevision);
+}, [propRevision]);
```

**Comment by markerikson** [comment]
Still an issue with v19 as far as I know.

**Comment by farwayer** [comment]
Hey, any news?

Just encountered this issue when the store change happens after getSnapshot is called during component rendering, but before subscribe is called. This bug is quite hard to debug.

It seems correct to not delay the subscribe call, but to do it right away inside useSyncExternalStore, otherwise you can lose changes.

**Comment by sophiebits** [comment]
Reproed the linked codesandbox has the same issue on 19.0.0-experimental-0bc30748-20241028.

**Comment by jquense** [comment]
Fairly certain we are hitting this as well but it's extremely hard to boil down into a shareable repro, even tho i can can reproduce this consistently in our app. So unclear if it's the same or different bug. 

From what I can tell, certain changes to the store don't ever get stamped on the `inst` so you change a  value from `null` to -> `'foo'`, the store responds to and triggers an update and runs `updateSyncExternalStore` but the pushed effect that runs `updateStoreInstance` never runs. This causes a problem if you then change the store value back to the original `null`, it's still got the first value to compare with, and sees "no update" and doesn't respond.

react 18.3
we are using suspense widely 
updates to the store don't trigger suspense but do trigger slow updates (a big data table changes)

**Comment by DanFromUkraine** [comment]
bump

**Comment by guille-stripe** [comment]
bump

**Comment by songkeys** [comment]
this still happen in the latest experimental build

**Comment by pllcoster** [comment]
I think we are experiencing an offshoot of this bug where if you are using Activity alongside React Compiler and (in our case) react-redux, you can sometimes be left with permanently stale UI, when an update to the redux store reveals components from a hidden activity tree, and that same redux store value is used by the (previously hidden) component in its UI.

The revealed component seems to be rendered from its last cached state and does not re-render with the updated state from the redux store, even though it has changed.

<img width="871" height="409" alt="Image" src="https://github.com/user-attachments/assets/f9a9c6d7-f1b4-4a92-8f80-ec2f08b37b44" />

See a reproduction case here: https://stackblitz.com/edit/vitejs-vite-avmvn96w?file=README.md
